File: 217. Contains Duplicate/solution.py
# What is a counter - A counter is a container included in the collections module
# What is a container - Containers are objects that hold objects. They provide a way to access the contained objects and iterate over them.
# A counter is a subclass of dict.



# read more about Counter here - https://www.geeksforgeeks.org/counters-in-python-set-1/
from collections import Counter

class Solution:
    def containsDuplicate(self, nums):

        # counter creates a k,v pair dictionary
        # where k is the element in the nums array
        # and value is the frequency of the key
        d = Counter(nums)
            
        # see how we use items for counter
        for key,value in d.items():
           # means the key has frequency greater than 1
           if value > 1:
                return True
            
        return False

File: 217. Contains Duplicate/solution.cpp
/** we have 2 solutions for cpp **/

// Time : O(n)
// Space : O(n)

// unordered_set - https://www.geeksforgeeks.org/unordered_set-in-cpp-stl/


// solution 1

#include <bits/stdc++.h> 
using namespace std;

class Solution {
	public:
		bool containsDuplicate(vector<int>& nums) {
			// initializing a new unordered_set from beginning of nums till the end
			unordered_set<int> nums_set(nums.begin(),nums.end());
			return nums_set.size() != nums.size();
		}
};

// Time : O(nlogn)
// Space : O(1)

class Solution {
	public: 
		bool containsDuplicate(vector<int> &nums) {
			// learn more about std::sort here - https://www.geeksforgeeks.org/sort-c-stl/
			sort(nums.begin(), nums.end());
			// read more about unique here - https://www.geeksforgeeks.org/stdunique-in-cpp/
			return unique(nums.begin(), nums.end()) != nums.end();
		}
};


